mail_sendgrid/models/mail_mail.py

```diff
@@ -0,0 +1,252 @@
+        Personalization, Substitution, Mail, Header
+except ImportError:
+    _logger.error("ImportError raised while loading module.")
+    _logger.debug("ImportError details:", exc_info=True)
+
+
+STATUS_OK = 202
+
+
+class MailMessage(models.Model):
+    """ Add SendGrid related fields so that they dispatch in all
+    subclasses of mail.message object
+    """
+    _inherit = 'mail.message'
+
+    ##########################################################################
+    #                                 FIELDS                                 #
+    ##########################################################################
+    body_text = fields.Text(help='Text only version of the body')
+    sent_date = fields.Datetime(copy=False)
+    substitution_ids = fields.Many2many(
+        'sendgrid.substitution', string='Substitutions', copy=True)
+    sendgrid_template_id = fields.Many2one(
+        'sendgrid.template', 'Sendgrid Template')
+    send_method = fields.Char(compute='_compute_send_method')
+
+    ##########################################################################
+    #                             FIELDS METHODS                             #
+    ##########################################################################
+    @api.multi
+    def _compute_send_method(self):
+        """ Check whether to use traditional send method, sendgrid or disable.
+        """
+        send_method = self.env['ir.config_parameter'].get_param(
+            'mail_sendgrid.send_method', 'traditional')
+        for email in self:
+            email.send_method = send_method
+
+
+class OdooMail(models.Model):
+    """ Email message sent through SendGrid """
+    _inherit = 'mail.mail'
+
+    ##########################################################################
+    #                                 FIELDS                                 #
+    ##########################################################################
+    tracking_email_ids = fields.One2many(
+        'mail.tracking.email', 'mail_id', string='Registered events',
+        readonly=True)
+    click_count = fields.Integer(
+        compute='_compute_tracking', store=True, readonly=True)
+    opened = fields.Boolean(
+        compute='_compute_tracking', store=True, readonly=True)
+    tracking_event_ids = fields.One2many(
+        'mail.tracking.event', compute='_compute_events')
+
+    @api.depends('tracking_email_ids', 'tracking_email_ids.click_count',
+                 'tracking_email_ids.state')
+    def _compute_tracking(self):
+        for email in self:
+            email.click_count = sum(email.tracking_email_ids.mapped(
+                'click_count'))
+            opened = len(email.tracking_email_ids.filtered(
+                lambda t: t.state == 'opened'))
+            email.opened = opened > 0
+
+    def _compute_events(self):
+        for email in self:
+            email.tracking_event_ids = email.tracking_email_ids.mapped(
```

Comment:
same story: if you expect many records use a search

Reply:
search instead of mapped? for better performance?